diff --git a/sapi/fpm/fpm/fpm_signals.c b/sapi/fpm/fpm/fpm_signals.c
index d3214b3d16..91b70e5713 100644
--- a/sapi/fpm/fpm/fpm_signals.c
+++ b/sapi/fpm/fpm/fpm_signals.c
@@ -184,7 +184,8 @@ static void sig_handler(int signo) /* {{{ */
 
 int fpm_signals_init_main() /* {{{ */
 {
-	struct sigaction act;
+	struct sigaction act, act_ign;
+	char* ignore_sigterm = getenv("HEROKU_PHP_GRACEFUL_SIGTERM");
 
 	if (0 > socketpair(AF_UNIX, SOCK_STREAM, 0, sp)) {
 		zlog(ZLOG_SYSERROR, "failed to init signals: socketpair()");
@@ -205,7 +206,10 @@ int fpm_signals_init_main() /* {{{ */
 	act.sa_handler = sig_handler;
 	sigfillset(&act.sa_mask);
 
-	if (0 > sigaction(SIGTERM,  &act, 0) ||
+	memset(&act_ign, 0, sizeof(act_ign));
+	act_ign.sa_handler = SIG_IGN;
+
+	if (0 > sigaction(SIGTERM,  (ignore_sigterm && atoi(ignore_sigterm)) ? &act_ign : &act,  0) ||
 	    0 > sigaction(SIGINT,   &act, 0) ||
 	    0 > sigaction(SIGUSR1,  &act, 0) ||
 	    0 > sigaction(SIGUSR2,  &act, 0) ||
@@ -226,20 +230,24 @@ int fpm_signals_init_main() /* {{{ */
 
 int fpm_signals_init_child() /* {{{ */
 {
-	struct sigaction act, act_dfl;
+	struct sigaction act, act_dfl, act_ign;
+	char* ignore_sigterm = getenv("HEROKU_PHP_GRACEFUL_SIGTERM");
 
 	memset(&act, 0, sizeof(act));
 	memset(&act_dfl, 0, sizeof(act_dfl));
+	memset(&act_ign, 0, sizeof(act_ign));
 
 	act.sa_handler = &sig_soft_quit;
 	act.sa_flags |= SA_RESTART;
 
 	act_dfl.sa_handler = SIG_DFL;
 
+	act_ign.sa_handler = SIG_IGN;
+
 	close(sp[0]);
 	close(sp[1]);
 
-	if (0 > sigaction(SIGTERM,  &act_dfl,  0) ||
+	if (0 > sigaction(SIGTERM,  (ignore_sigterm && atoi(ignore_sigterm)) ? &act_ign : &act_dfl,  0) ||
 	    0 > sigaction(SIGINT,   &act_dfl,  0) ||
 	    0 > sigaction(SIGUSR1,  &act_dfl,  0) ||
 	    0 > sigaction(SIGUSR2,  &act_dfl,  0) ||
